File: golang/baseweb/context.go
package baseweb

import (
	"context"
	"fmt"
	"net/http"
)

// Context 上下文
type Context struct {
	request  *http.Request
	response http.ResponseWriter

	context.Context
	URI         string
	Method      string
	Params      map[string]string
	QueryParams map[string]string
	middleware  []Handler
	nextIdx     int
	StatusCode  int
}

// NewContext 创建上下文
func NewContext(res http.ResponseWriter, req *http.Request) *Context {
	return &Context{
		request:     req,
		response:    res,
		Context:     context.Background(),
		URI:         req.URL.Path,
		Method:      req.Method,
		StatusCode:  200,
		Params:      make(map[string]string),
		QueryParams: make(map[string]string),
		middleware:  make([]Handler, 0),
		nextIdx:     -1,
	}
}

// Next 下一步
func (c *Context) Next() {
	c.nextIdx++
	if c.nextIdx >= len(c.middleware) {
		return
	}
	c.middleware[c.nextIdx](c)
}

// Send 发送响应
func (c *Context) Send(data string) {
	c.response.WriteHeader(c.StatusCode)
	fmt.Fprintf(c.response, data)
}

File: dart/lib/src/context.dart
import 'dart:io';

import 'router.dart';

class Context {
  String uri;
  String method;

  Map<String, String> params;
  Map<String, String> queryParams;

  List<Handler> handlers;
  int statusCode;
  HttpRequest request;
  HttpResponse response;
  int _nextIdx = -1;

  Context(HttpRequest req, HttpResponse res) {
    this.uri = req.uri.path;
    this.method = req.method;
    this.request = req;
    this.response = res;
    this.statusCode = 200;
    this.handlers = [];
    this.params = new Map<String, String>();
    this.queryParams = new Map<String, String>();
  }

  next() async {
    this._nextIdx++;
    if (this._nextIdx >= this.handlers.length) return;

    try {
      await this.handlers[this._nextIdx](this);
    } catch (e) {
      print(e);
      this.abort();
    }
  }

  abort() {
    this._nextIdx = this.handlers.length;
  }

  send(String data) {
    this.response.statusCode = this.statusCode ?? 200;
    this.response.write(data);
    this.response.close();
  }
}

